### PR TITLE
tag qol

### PR DESCRIPTION
**Pull request by illuminator3**
image

### CODE REVIEW

**Comment by Zabuzard** [comment]
Im confused. What is this? Is that for #367 ? If so, link it please.

**Comment by Taz03** [comment]
> Im confused. What is this? Is that for #594 ? If so, link it please. Also, that is already assigned to @Taz03 .

its a little different, the auto-complete provides option before the command usage, but lets say the user still writes and invalid id (which is possible) then the reply will contain some buttons for possible suggestions

**Comment by Zabuzard** [comment]
i see. fair.

File: application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java [hunk 1]
```diff
@@ -104,12 +104,12 @@ public BotCore(JDA jda, Database database, Config config) {
...
         Collection<UserInteractor> interactors = getInteractors();
 
-        interactors.forEach(slashCommand -> slashCommand
-            .acceptComponentIdGenerator(((componentId, lifespan) -> {
-                UUID uuid = UUID.randomUUID();
-                componentIdStore.putOrThrow(uuid, componentId, lifespan);
-                return uuid.toString();
-            })));
+        interactors.forEach(
+                interactor -> interactor.acceptComponentIdGenerator(((componentId, lifespan) -> {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java:108]: please rebase.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 2]
```diff
@@ -2,23 +2,38 @@
...
 import java.util.stream.Collectors;
 
 /**
  * The core of the tag system. Provides methods to read and create tags, directly tied to the
  * underlying database.
  */
 public final class TagSystem {
+    /**
+     * The amount of candidates that are suggested as buttons in the message that tells the user
+     * that the tag they tried accessing doesn't exist
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:28]: this is too specific and implementation-dependent. if someone changes how the suggestions are down, they will definitely forget to update this text and it rots.

maybe change to "The amount of tags suggested when the user types an incorrect tag name."

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 3]
```diff
@@ -2,23 +2,38 @@
...
+     */
+    private static final int CANDIDATE_SUGGESTIONS_PER_USAGE = 5;
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:30]: I dont like the name of this. maybe `TAG_SUGGESTIONS_AMOUNT`

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 4]
```diff
@@ -2,23 +2,38 @@
...
+
+    /**
+     * todo document
+     */
+    static final String TAG_SUGGESTION_INDICATOR = "tag suggestion";
+
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:36]: what is this? also, still has TODO in it

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 5]
```diff
@@ -57,17 +72,40 @@ static Button createDeleteButton(String componentId) {
      * @param event the event to send messages with
      * @return whether the given tag is unknown to the system
      */
-    boolean handleIsUnknownTag(String id, IReplyCallback event) {
+    boolean handleIsUnknownTag(String id, IReplyCallback event,
+            Function<String[], String> componentIdGenerator,
+            @Nullable OptionMapping userToReplyTo) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:77]: what is this? also, missing documentation. ill recheck ur code for this once u added doc.

File: application/src/main/java/org/togetherjava/tjbot/commands/BotCommandAdapter.java [hunk 6]
```diff
@@ -130,7 +130,8 @@ protected final String generateComponentId(String... args) {
      */
     @SuppressWarnings({"OverloadedVarargsMethod", "WeakerAccess"})
     protected final String generateComponentId(Lifespan lifespan, String... args) {
-        return componentIdGenerator.generate(new ComponentId(getName(), Arrays.asList(args)),
-                lifespan);
+        return componentIdGenerator
+            .generate(new ComponentId(UserInteractorPrefix.getPrefixedNameFromInstance(this),
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/BotCommandAdapter.java:134]: please rebase.

File: application/src/test/java/org/togetherjava/tjbot/commands/SlashCommandAdapterTest.java [hunk 7]
```diff
@@ -10,6 +10,8 @@
 
 final class SlashCommandAdapterTest {
     private static final String NAME = "foo";
+    private static final String PREFIXED_NAME =
```
Comment (Zabuzard) [inline application/src/test/java/org/togetherjava/tjbot/commands/SlashCommandAdapterTest.java:13]: please rebase

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 8]
```diff
@@ -46,20 +50,40 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
...
-        message.queue();
+
+        sendTagReply(event, event.getUser().getName(), args.get(1), Optional.empty(),
+                Optional.ofNullable(args.get(2)));
+    }
+
+    /**
+     * Sends the reply for a successfull /tag use (i.e. the given tag exists)
+     */
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:77]: why suppress? dont do it, use `@Nullable` instead.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 9]
```diff
@@ -46,20 +50,40 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
...
+                Optional.ofNullable(replyToUserOption)
+                    .map(OptionMapping::getAsUser)
+                    .map(IMentionable::getAsMention));
+    }
 
-        if (replyToUserOption != null) {
-            message = message.setContent(replyToUserOption.getAsUser().getAsMention());
+    @Override
+    public void onButtonClick(ButtonInteractionEvent event, List<String> args) {
+        if (!TagSystem.TAG_SUGGESTION_INDICATOR.equals(args.get(0))) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:66]: thats really dirty. also, nothing in the code prevents someone from creating a tag with that name and then everything goes bonkers.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 8 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:77]: then I'd have to use a bunch of ternaries inside which would look really ugly

File: application/src/test/java/org/togetherjava/tjbot/commands/SlashCommandAdapterTest.java (hunk 7 above)
Comment (illuminator3) [reply application/src/test/java/org/togetherjava/tjbot/commands/SlashCommandAdapterTest.java]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/BotCommandAdapter.java (hunk 6 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/BotCommandAdapter.java]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java (hunk 1 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 9 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:66]: how would I do it instead then?

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java (hunk 3 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:30]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 9 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:66]: maybe i just misunderstood what this does. why do u need it in the first place? this argument seems completely obsolete to me

Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:66]: to know if the button comes from a suggestion. otherwise it'd throw lots of errors if a button was clicked that doesn't come from us because the arguments are wrong. or am I misunderstanding when onButtonClick gets invoked?

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:66]: > or am I misunderstanding when onButtonClick gets invoked?

yes. its only invoked for buttons created by ur class. ur safety check is obsolete. `BotCore` already takes care of this for u.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java (hunk 4 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:36]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 8 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:77]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 9 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:66]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java (hunk 5 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:76]: recheck please

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 10]
```diff
@@ -46,20 +50,35 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
...
+    @Override
+    public void onButtonClick(ButtonInteractionEvent event, List<String> args) {
+        sendTagReply(event, event.getUser().getName(), args.get(0), null, args.get(1));
+    }
+
+    /**
+     * Sends the reply for a successfull /tag use (i.e. the given tag exists)
+     */
+    private void sendTagReply(IReplyCallback callback, String userName, String tag,
+            @Nullable String commandString, @Nullable String replyToUser) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:71]: I have some trouble with some variable names. Some of them dont really say what they are about.

* `callback` - callback of what? I think it would be easier and more idiomatic if you just call it `event` (even if the type is a level higher).
* its easy to confuse the two users that play a role here, since both are called "user". maybe keep `userName`, but rename `replyToUser` to `replyTargetUser`. if u want, rename `userName` to `invokerUserName` maybe, then its super clear.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 11]
```diff
@@ -46,20 +50,35 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
...
                 .setTimestamp(Instant.now())
                 .setColor(TagSystem.AMBIENT_COLOR)
-                .build());
-
-        if (replyToUserOption != null) {
-            message = message.setContent(replyToUserOption.getAsUser().getAsMention());
-        }
-        message.queue();
+                .build())
+            .setContent(replyToUserOpt.orElse(""))
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:81]: dont know the javadoc of `setContent`, but many JDA methods accept `null`. If thats the case, this could be a nice improvement, since u dont need the optional anymore then

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagManageCommand.java [hunk 12]
```diff
@@ -152,7 +152,7 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
 
     private void rawTag(SlashCommandInteractionEvent event) {
         String id = Objects.requireNonNull(event.getOption(ID_OPTION)).getAsString();
-        if (tagSystem.handleIsUnknownTag(id, event)) {
+        if (tagSystem.handleIsUnknownTag(id, event, super.getComponentIdGenerator(), null)) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagManageCommand.java:155]: NIT: the `super` is unecessary

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagManageCommand.java [hunk 13]
```diff
@@ -317,7 +317,7 @@ private Optional<String> getTagContent(Subcommand subcommand, String id) {
     private boolean isWrongTagStatusAndHandle(TagStatus requiredTagStatus, String id,
             IReplyCallback event) {
         if (requiredTagStatus == TagStatus.EXISTS) {
-            return tagSystem.handleIsUnknownTag(id, event);
+            return tagSystem.handleIsUnknownTag(id, event, super.getComponentIdGenerator(), null);
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagManageCommand.java:320]: NIT: the `super` is unecessary

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 14]
```diff
@@ -54,19 +67,41 @@ static Button createDeleteButton(String componentId) {
      *
      * @param id the id of the tag to check
      * @param event the event to send messages with
+     * @param componentIdGenerator the component id generator to generate the button ids with
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:70]: what buttons? could help to give some context to the javadoc. such as `@param componentIdGenerator used to generate buttons with tag suggestions`

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 15]
```diff
@@ -54,19 +67,41 @@ static Button createDeleteButton(String componentId) {
...
+     * @param userToReplyTo the user that was originally meant to be replied to when the tag command
+     *        was invoked
      * @return whether the given tag is unknown to the system
      */
-    boolean handleIsUnknownTag(String id, IReplyCallback event) {
+    boolean handleIsUnknownTag(String id, IReplyCallback event,
+            ComponentIdGenerator componentIdGenerator, @Nullable OptionMapping userToReplyTo) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:76]: NIT: same naming issue with the reply user. maybe `replyTargetUser`. But `userToReplyTo` is also okay, just sounds a bit wordy.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 16]
```diff
@@ -54,19 +67,41 @@ static Button createDeleteButton(String componentId) {
...
-            .orElse(".");
 
-        event.reply("Could not find any tag with id '%s'%s".formatted(id, suggestionText))
-            .setEphemeral(true)
-            .queue();
+        List<String> candidates = getAllIds().stream()
+            .sorted(Comparator
+                .comparingInt(candidate -> StringDistances.editDistance(id, candidate)))
+            .limit(TAG_SUGGESTIONS_AMOUNT)
+            .toList();
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:85]: Id suggest using a PQ with k-sort instead. Would be faster if the amount of tags increases further.

```java
Queue<String> closestMatches = new PriorityQueue<>(Comparator...);
closestMatches.addAll(getAllIds());

List<String> suggestions = Stream.generate(closestMatches::poll)
  .limit(TAG_SUGGESTIONS_AMOUNT)
  .toList();
```

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 17]
```diff
@@ -54,19 +67,41 @@ static Button createDeleteButton(String componentId) {
...
+        List<List<String>> partition = ListUtils.partition(candidates, 5);
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:86]: move that down a few lines, to where its used first

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 18]
```diff
@@ -54,19 +67,41 @@ static Button createDeleteButton(String componentId) {
...
+        ReplyCallbackAction action =
+                event
+                    .reply("Could not find any tag with id '%s'%s".formatted(id,
+                            candidates.isEmpty() ? "."
+                                    : ", did you perhaps mean any of the following?"))
+                    .setEphemeral(true);
+
+        for (List<String> part : partition) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:94]: naming issue: `partition` and `partitions`, would be easier to understand.

Or `batch` and `batches`

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 19]
```diff
@@ -54,19 +67,41 @@ static Button createDeleteButton(String componentId) {
...
+            action.addActionRow(part.stream()
+                .map(i -> Button.secondary(componentIdGenerator.generate(new ComponentId(
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:96]: `i` is confusing, sounds like an index, but its actually the `suggestion`. please write it out

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 20]
```diff
@@ -54,19 +67,41 @@ static Button createDeleteButton(String componentId) {
...
+                        UserInteractorPrefix.getPrefixedNameFromClass(TagCommand.class, "tag"),
+                        Arrays.asList(i,
+                                userToReplyTo != null ? userToReplyTo.getAsUser().getAsMention()
+                                        : null)),
+                        Lifespan.REGULAR), i))
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:101]: This whole thing reads a bit complex, being nested in the `map` and loop. Id move this button creation out to a quick private helper method. sth like `map(this::createSuggestionButton)` or similar.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java [hunk 21]
```diff
@@ -54,19 +67,41 @@ static Button createDeleteButton(String componentId) {
...
+                event
+                    .reply("Could not find any tag with id '%s'%s".formatted(id,
+                            candidates.isEmpty() ? "."
+                                    : ", did you perhaps mean any of the following?"))
+                    .setEphemeral(true);
+
+        for (List<String> part : partition) {
+            action.addActionRow(part.stream()
+                .map(i -> Button.secondary(componentIdGenerator.generate(new ComponentId(
+                        UserInteractorPrefix.getPrefixedNameFromClass(TagCommand.class, "tag"),
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagSystem.java:97]: the usage of `"tag"` here is a bit dangerous. if someone renames the command, they will for sure forget to adjust it here as well.

try to link it somehow. either by giving it as method argument (like u did with the component ID genererator), or by having a constant between here and `TagCommand`.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/BotCommandAdapter.java [hunk 22]
```diff
@@ -136,10 +137,13 @@ protected final String generateComponentId(Lifespan lifespan, String... args) {
...
-     * @return the component id generator used by {@link #generateComponentId(String...)} and
-     *         {@link #generateComponentId(Lifespan, String...)}
+     * Gets the generator used to create component IDs.
+     * <p>
+     * In general, prefer using {@link #generateComponentId(Lifespan, String...)} and {@link #generateComponentId(String...)} instead of interacting with the generator directly.
+     *
+     * @return the generator
      */
-    protected final ComponentIdGenerator getComponentIdGenerator() {
+    protected final @Nullable ComponentIdGenerator getComponentIdGenerator() {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/BotCommandAdapter.java:147]: the method is **not** supposed to return a nullable.

its a mistake to call the method before intiialization, not allowed. the method should fail-fast if its called before.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

**Comment by Taz03** [comment]
this pr is not valid anymore, we use auto-complete feature now, and while using this feature a user can only put the registered option, u can close this pr now

**Comment by illuminator3** [comment]
thank you for appreciating all of the work I put into this already

**Comment by Tais993** [comment]
> this pr is not valid anymore, we use auto-complete feature now, and while using this feature a user can only put the registered option, u can close this pr now

This is false, the user can still put any text that they want, autocomplete doesn't force anything.

**Comment by ankitsmt211** [comment]
This PR is being closed due to inactivity, can be opened later if need be.
